### PR TITLE
Added NLA detection and redirection

### PR DESCRIPTION
**Pull request by Pourliver**
As discussed, I added a way to scan the RDP's configuration of our target whenever someone connects to him. If that target has NLA activated, there are two possible scenario : 

-    Redirect the client to a non-NLA host that we configure as a parameter. This has been done, and it's impossible for the client to detect that we forwarded him somewhere else.
-    Relay the client to another service with mitmrelayx from Impacket. This hasn't been done since I need to talk with @Romounet first.

Overall, this works really well. During my test, I wasn't even aware I've been redirected since Windows 10's RDP client was still showing my target's IP in the top bar, even on the lock screen. I could only tell once I authenticated.

I did a few caplets to simplify these payloads, I will link them here.

I also refactored the handler's logic a bit to make it more readable as well. Let me know if anything's wrong.

EDIT : Caplets : https://github.com/GoSecure/caplets/tree/master/rdp-proxy

### CODE REVIEW

File: modules/rdp_proxy/rdp_proxy_linux_amd64.go [hunk 1]
```diff
@@ -61,13 +70,20 @@ func NewRdpProxy(s *session.Session) *RdpProxy {
...
 mod.AddParam(session.NewStringParameter("rdp.proxy.out", "./", "", "The output directory for PyRDP artifacts."))
 mod.AddParam(session.NewStringParameter("rdp.proxy.targets", session.ParamSubnet, "", "Comma separated list of IP addresses to proxy to, also supports nmap style IP ranges."))
 mod.AddParam(session.NewStringParameter("rdp.proxy.regexp", "(?i)(cookie:|mstshash=|clipboard data|client info|credential|username|password|error)", "", "Print PyRDP logs matching this regular expression."))
+// Optional paramaters
+mod.AddParam(session.NewStringParameter("rdp.proxy.nla.seccheck", "", "", "Path to rdp-sec-check.pl. Allows more complex exploits when NLA is enforced (optional)."))
+mod.AddParam(session.NewStringParameter("rdp.proxy.nla.mode", "IGNORE", "(IGNORE|RELAY|REDIRECT)", "Specify how to handle connections to a NLA-enabled host. Require rdp.proxy.nla.seccheck."))
+mod.AddParam(session.NewStringParameter("rdp.proxy.nla.redirectip", "", "", "Specify IP to redirect clients that connects to NLA targets. Require rdp.proxy.nla.mode REDIRECT"))
```
Comment (alxbl) [inline modules/rdp_proxy/rdp_proxy_linux_amd64.go]: I would make this `rdp.proxy.nla.redirect.ip`

File: modules/rdp_proxy/rdp_proxy_linux_amd64.go [hunk 2]
```diff
@@ -61,13 +70,20 @@ func NewRdpProxy(s *session.Session) *RdpProxy {
...
+mod.AddParam(session.NewIntParameter("rdp.proxy.nla.redirectport", "3389", "Specify port to redirect clients that connects to NLA targets. Require rdp.proxy.nla.mode REDIRECT"))
```
Comment (alxbl) [inline modules/rdp_proxy/rdp_proxy_linux_amd64.go]: `redirect.port`

File: modules/rdp_proxy/rdp_proxy_linux_amd64.go [hunk 3]
```diff
@@ -93,6 +109,52 @@ func (mod *RdpProxy) isTarget(ip string) bool {
     return false
 }
 
+func (mod *RdpProxy) isNLAEnforced(target string) (nla bool, err error) {
+    if mod.secCheck != "" {
```
Comment (alxbl) [inline modules/rdp_proxy/rdp_proxy_linux_amd64.go]: You could invert the `if` condition to make the code less nested

File: modules/rdp_proxy/rdp_proxy_linux_amd64.go [hunk 4]
```diff
@@ -113,26 +175,26 @@ func (mod *RdpProxy) filterLogs(prefix string, output io.ReadCloser) {
 }
 
 // Adds the firewall rule for proxy instance.
-func (mod *RdpProxy) doProxy(addr string, port string, enable bool) (err error) {
+func (mod *RdpProxy) doProxy(dst string, proxyPort string) (err error) {
```
Comment (alxbl) [inline modules/rdp_proxy/rdp_proxy_linux_amd64.go:216]: Thanks for fixing this!

File: modules/rdp_proxy/rdp_proxy_linux_amd64.go [hunk 5]
```diff
@@ -240,53 +314,70 @@ func (mod *RdpProxy) Configure() (err error) {
...
-                payload.SetVerdict(nfqueue.NF_DROP)
-
-                return 0
+            targetNLA, _ := mod.isNLAEnforced(target)
+
+            // Only if seccheck is set
+            if targetNLA {
+                switch mod.nlaMode {
+                case "RELAY":
+                    mod.Info("%s Target has NLA enabled and mode RELAY, unimplemented", ips)
```
Comment (alxbl) [inline modules/rdp_proxy/rdp_proxy_linux_amd64.go]: Maybe make this check at plugin startup (when parsing options) before letting the user run in relay mode

**Review by alxbl — COMMENTED** [review]
This is looking good so far.

I don't like adding a lot of external dependencies like that as it makes the module hard to use, but I'm fine with it for now as a proof of concept. Besides, we already depend on pyrdp being present...

The only thing I'd suggest is documenting the required external binaries somewhere.

**Comment by Pourliver** [comment]
I agree, I am not a fan of the dependency either. This was only for a PoC. I'll look a bit into the code and see if I can easily extract the needed packets to remove this dependency. Both project licenses aren't compatible sadly (GPL-2.0 and GPL-3.0)

**Comment by Pourliver** [comment]
I took the day yesterday to look further into rdp-sec-check. It was easy enough to capture the TCP payload that validated NLA. I hard-coded those payloads inside rdp-proxy, and removed the dependency.

**Review by alxbl — APPROVED** [review]
Very nice!

**Pull request merged**